src/op_analytics/cli/subcommands/pulls/defillama.py

```diff
@@ -88,60 +83,111 @@ def process_breakdown_stables(data):
 
+    if "peggedAssets" not in summary:
+        raise KeyError("The 'peggedAssets' key is missing from the summary data.")
+
     urls = {}
     for stablecoin in summary["peggedAssets"]:
         stablecoin_id = stablecoin["id"]
-        urls[stablecoin_id] = BREAKDOWN_ENDPOINT.format(id=stablecoin_id)
+        if stablecoin_ids is None or stablecoin_id in stablecoin_ids:
+            urls[stablecoin_id] = BREAKDOWN_ENDPOINT.format(id=stablecoin_id)
+    if not urls:
+        raise ValueError("No valid stablecoin IDs provided.")
 
-    stablecoin_data = run_concurrently(lambda x: get_data(session, x), urls, max_workers=4)
+    stablecoin_data = run_concurrently(
+        lambda x: get_data(session, x), urls, max_workers=4
+    )
 
-    breakdown_dfs = []
-    metadata_rows = []
-    for data in stablecoin_data.values():
-        breakdown_df, metadata = process_breakdown_stables(data)
-        breakdown_dfs.append(breakdown_df)
-        metadata_rows.append(metadata)
+    breakdown_dfs: List[pl.DataFrame] = []
+    metadata_rows: List[Dict[str, Optional[str]]] = []
 
-    breakdown_df = pl.concat(breakdown_dfs, how="diagonal_relaxed")
-    metadata_df = pl.DataFrame(metadata_rows, infer_schema_length=len(metadata_rows))
+    for data in stablecoin_data.values():
+        breakdown_df, metadata = process_breakdown_stables(data, days=days)
+        if not breakdown_df.is_empty():
+            breakdown_dfs.append(breakdown_df)
+            if metadata:
+                metadata_rows.append(metadata)
+
+    breakdown_df = (
+        pl.concat(breakdown_dfs, how="diagonal_relaxed")
+        if breakdown_dfs
+        else pl.DataFrame()
+    )
+    metadata_df = (
+        pl.DataFrame(metadata_rows, infer_schema_length=len(metadata_rows))
+        if metadata_rows
+        else pl.DataFrame()
+    )
 
-    # Write metadata to BQ
     dt = now_date()
-    overwrite_unpartitioned_table(metadata_df, BQ_DATASET, f"{METADATA_TABLE}_latest")
-    overwrite_partition_static(metadata_df, dt, BQ_DATASET, f"{METADATA_TABLE}_history")
 
-    # Write breakdown to BQ. Only the most recent 7 days are inserted.
-    # Data in BQ older than 7 days is considered immutable.
-    overwrite_partitions_dynamic(
-        df=most_recent_dates(breakdown_df, n_dates=7),
+    if metadata_df.is_empty():
+        raise ValueError(
+            "metadata_df is empty. Expected non-empty data to write to BigQuery."
+        )
+    overwrite_unpartitioned_table(metadata_df, BQ_DATASET, f"{METADATA_TABLE}_latest")
+    overwrite_partition_static(
+        metadata_df,
+        partition_dt=dt,
         dataset=BQ_DATASET,
-        table_name=f"{BREAKDOWN_TABLE}_history",
+        table_name=f"{METADATA_TABLE}_history",
     )
 
+    if breakdown_df.is_empty():
+        raise ValueError(
+            "breakdown_df is empty. Expected non-empty data to write to BigQuery."
+        )
+
+    dates = breakdown_df["dt"].unique().to_list()
```

Comment:
I think that the functionality of splitting the data frame into `dt` parts and making sure each partition is upserted accordingly should be the responsibility of the `upsert_partitioned_table` function.  This is almost certainly going to come up again each time we want to use `upsert_partitioned_table`.